File: Data Structure & Algorithm/Data Structure/Graph/Disjoint Set/Theory/QuickUnion.py
# UnionFind class
# Time Complexity	O(α(N))

# Note:
# N is the number of vertices in the graph.
# α refers to the Inverse Ackermann function.
# In practice, we assume it's a constant. In other words, O(α(N)) is regarded as O(1) on average.

# Space Complexity O(N)

class UnionFind:
    def __init__(self, size):
        self.root = [i for i in range(size)]
        # Use a rank array to record the height of each vertex, i.e., the "rank" of each vertex.
        # The initial "rank" of each vertex is 1, because each of them is
        # a standalone vertex with no connection to other vertices.
        self.rank = [1] * size
        self.count = size

    # The find function here is the same as that in the disjoint set with path compression.
    def find(self, x):
        if x == self.root[x]:
            return x
        self.root[x] = self.find(self.root[x])
        return self.root[x]

    # The union function with union by rank
    def union(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX != rootY:
            if self.rank[rootX] > self.rank[rootY]:
                self.root[rootY] = rootX
            elif self.rank[rootX] < self.rank[rootY]:
                self.root[rootX] = rootY
            else:
                self.root[rootY] = rootX
                self.rank[rootX] += 1
            self.count = self.count - 1

    def connected(self, x, y):
        return self.find(x) == self.find(y)
    
    def getRootCount(self):
        return self.count


# Test Case
uf = UnionFind(10)
# 1-2-5-6-7 3-8-9 4
uf.union(1, 2)
uf.union(2, 5)
uf.union(5, 6)
uf.union(6, 7)
uf.union(3, 8)
uf.union(8, 9)
print(uf.connected(1, 5))  # true
print(uf.connected(5, 7))  # true
print(uf.connected(4, 9))  # false
# 1-2-5-6-7 3-8-9-4
uf.union(9, 4)
print(uf.connected(4, 9))  # true

File: Data Structure & Algorithm/Data Structure & Algorithm [Old]/Models/Graph/Quick Union/Ranking/QuickUnion.go
// Time Complexity	O(α(N))

// Note:
// N is the number of vertices in the graph.
// α refers to the Inverse Ackermann function.
// In practice, we assume it's a constant. In other words, O(α(N)) is regarded as O(1) on average.

// Space Complexity O(N)

package main

import "fmt"

// UnionFind structure
type UnionFind struct {
	root []int
	rank []int
}

// NewUnionFind initializes a new UnionFind structure
func NewUnionFind(size int) *UnionFind {
	root := make([]int, size)
	rank := make([]int, size)
	for i := range root {
		root[i] = i
		rank[i] = 1
	}
	return &UnionFind{root: root, rank: rank}
}

// Find finds the root of x with path compression
func (uf *UnionFind) Find(x int) int {
	if uf.root[x] != x {
		uf.root[x] = uf.Find(uf.root[x])
	}
	return uf.root[x]
}

// Union unions two sets
func (uf *UnionFind) Union(x, y int) {
	rootX := uf.Find(x)
	rootY := uf.Find(y)
	if rootX != rootY {
		if uf.rank[rootX] > uf.rank[rootY] {
			uf.root[rootY] = rootX
		} else if uf.rank[rootX] < uf.rank[rootY] {
			uf.root[rootX] = rootY
		} else {
			uf.root[rootY] = rootX
			uf.rank[rootX]++
		}
	}
}

// Connected checks if two nodes are in the same set
func (uf *UnionFind) Connected(x, y int) bool {
	return uf.Find(x) == uf.Find(y)
}

func main() {
	uf := NewUnionFind(10)
	uf.Union(1, 2)
	uf.Union(2, 3)
	uf.Union(4, 5)
	uf.Union(6, 7)
	uf.Union(8, 9)
	uf.Union(1, 9)
	fmt.Println(uf.Connected(1, 9))
	fmt.Println(uf.Connected(3, 9))
	fmt.Println(uf.Connected(1, 4))
	fmt.Println(uf.Connected(8, 4))
	fmt.Println(uf.Connected(4, 5))
	fmt.Println(uf.Connected(5, 6))
	fmt.Println(uf.Connected(6, 7))
}